src/Data/Tree/AVL/Indexed/Relation/Unary/Any/Properties.agda

```diff
@@ -0,0 +1,180 @@
+                (r : ∃ λ i → Tree V [ kv .key ] u (i ⊕ hʳ)) →
+                (bal : hˡ ∼ hʳ ⊔ h) →
+                Any P (proj₂ r) → Any P (proj₂ (joinʳ⁺ kv l r bal))
+joinʳ⁺-right⁺ k₂ t₁ (0# , t₃)                       bal p                = right p
+joinʳ⁺-right⁺ k₂ t₁ (1# , t₃)                       ∼0  p                = right p
+joinʳ⁺-right⁺ k₂ t₁ (1# , t₃)                       ∼-  p                = right p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼+)         ∼+  (here p)         = here p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼+)         ∼+  (left p)         = left (right p)
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼+)         ∼+  (right p)        = right p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼0)         ∼+  (here p)         = here p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼0)         ∼+  (left p)         = left (right p)
+joinʳ⁺-right⁺ k₂ t₁ (1# , node k₄ t₃ t₅ ∼0)         ∼+  (right p)        = right p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  (here p)         = right (here p)
+joinʳ⁺-right⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  (left (here p))  = here p
+joinʳ⁺-right⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  (left (left p))  = left (right p)
+joinʳ⁺-right⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  (left (right p)) = right (left p)
+joinʳ⁺-right⁺ k₂ t₁ (1# , node⁻ k₆ k₄ t₃ t₅ bal t₇) ∼+  (right p)        = right (right p)
+
+------------------------------------------------------------------------
+-- insert
+
+module _ {V : Value v} where
+
+  private Val = Value.family V
+
+  Any-insertWith :
+    ∀ {l u h} (k : Key) →
+    (f : Maybe (Val k) → Val k) →
+    (Pf : ∀ {k′} → (eq : k ≈ k′) (mk′ : Maybe (Val k′)) →
+          maybe′ (λ v → P (k′ , Value.respects V eq (f (just (Value.respects V (sym eq) v)))))
+                 (P (k , f nothing)) mk′)
+    (t : Tree V l u h) (seg : l < k < u) →
+    Any P (proj₂ (insertWith k f t seg))
+  Any-insertWith k f Pf (leaf l<u) l<k<u = here (Pf refl nothing)
+  Any-insertWith k f Pf (node kv@(k′ , v′) lp pu bal) (l<k , k<u) with compare k k′
+  ... | tri< k<k′ _ _ = joinˡ⁺-left⁺ kv (insertWith k f lp _) pu bal (Any-insertWith k f Pf lp (l<k , [ k<k′ ]ᴿ))
+  ... | tri> _ _ k′<k = joinʳ⁺-right⁺ kv lp (insertWith k f pu _) bal (Any-insertWith k f Pf pu ([ k′<k ]ᴿ , k<u))
+  ... | tri≈ _ k≈k′ _ = here (Pf k≈k′ (just v′))
+
+  Any-insert :
+    (P≈ : ∀ {k k′} v (eq : k ≈ k′) → P (k , v) → P (k′ , Value.respects V eq v)) →
+    ∀ {l u h} (k : Key) (v : Val k) → (Pv : P (k , v))
+    (t : Tree V l u h) (seg : l < k < u) →
+    Any P (proj₂ (insert k v t seg))
+  Any-insert P≈ k v Pv = Any-insertWith k (F.const v) $ λ where
+    eq nothing   → Pv
+    eq (just v′) → P≈ v eq Pv
+
+  insert⁺ : ∀ {l u h} (k : Key) (f : Maybe (Val k) → Val k) (t : Tree V l u h) (seg : l < k < u) →
+     (p : Any P t) → k ≉ Any.lookup p → Any P (proj₂ (insertWith k f t seg))
+  insert⁺ k f (node (k′ , v′) l r bal) (l<k , k<u) p k≉ with compare k k′
```

Comment:
The trick I usually use here is to write `with compare k k′ | p` which allows you to use ellipses to greatly reduce the size of the proof. Unsure what the computational implications are, but it does massively increase readability.